staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go

```diff
@@ -314,6 +379,59 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 					defer proxy.Close()
 				}
 
+				if testCase.socks5ProxyServerFunc != nil {
+					var l net.Listener
+					var proxyHandler *socks5.Server
+					if testCase.proxyAuth != nil {
+						proxyHandler = testCase.socks5ProxyServerFunc(&socks5.StaticCredentials{
+							"proxyuser": "proxypasswd",
+						})
+					} else {
+						proxyHandler = socks5Server(nil)
+					}
+
+					closed := false
+					l, err = net.Listen("tcp", "127.0.0.1:0")
+					if err != nil {
+						t.Fatalf("socks5Server: proxy_test: Listen: %v", err)
+					}
+					defer func() {
+						closed = true
+						l.Close()
+					}()
+
+					proxyCalledWithHost = "//127.0.0.1:" + strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
```

Comment:
the point of this var is to capture the host the proxy handler saw, to make sure 1) the proxy was contacted and 2) the proxy was asked to proxy to the right place.

the 
```
if testCase.proxyServerFunc != nil {
```
check below needs to be expanded to 
```
if testCase.proxyServerFunc != nil || testCase.socks5ProxyServerFunc != nil {
```

Once that is done, the check will start failing because this is being hard-coded to the proxy's host, not the backend's host. Letting interceptor set proxyCalledWithHost would match how this was intended to be used.

This diff would make the socks5 tests work the same way as the http proxy ones (and change each iteration of the loop to its own isolated test so you could do something like:
```
go test ./vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy -run TestRoundTripAndNewConnection//SOCKS5`
```

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
index 55aa5f30807..8cb0e2de476 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go
@@ -36,13 +36,16 @@ import (
 	"github.com/elazarl/goproxy"
 
 	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 type Interceptor struct {
 	Authorization       socks5.AuthContext
+	proxyCalledWithHost *string
 }
 
 func (i *Interceptor) Rewrite(ctx context.Context, req *socks5.Request) (context.Context, *socks5.AddrSpec) {
+	*i.proxyCalledWithHost = req.DestAddr.Address()
 	i.Authorization = socks5.AuthContext(*req.AuthContext)
 	return ctx, req.DestAddr
 }
@@ -86,8 +89,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 				return ts
 			}
 
-			interceptor := &Interceptor{}
-			socks5Server := func(creds *socks5.StaticCredentials) *socks5.Server {
+			socks5Server := func(creds *socks5.StaticCredentials, interceptor *Interceptor) *socks5.Server {
 				var conf *socks5.Config
 				if creds != nil {
 					cator := socks5.UserPassAuthenticator{Credentials: creds}
@@ -96,7 +98,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						Rewriter:    interceptor,
 					}
 				} else {
-					conf = &socks5.Config{}
+					conf = &socks5.Config{Rewriter: interceptor}
 				}
 
 				ts, err := socks5.New(conf)
@@ -108,7 +110,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 
 			testCases := map[string]struct {
 				serverFunc             func(http.Handler) *httptest.Server
-				socks5ProxyServerFunc  func(*socks5.StaticCredentials) *socks5.Server
+				socks5ProxyServerFunc  func(*socks5.StaticCredentials, *Interceptor) *socks5.Server
 				proxyServerFunc        func(http.Handler) *httptest.Server
 				proxyAuth              *url.Userinfo
 				clientTLS              *tls.Config
@@ -306,8 +308,9 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 				},
 			}
 
-			for k, testCase := range testCases {
-				testCase := testCase
+			for _, k := range sets.StringKeySet(testCases).List() {
+				testCase := testCases[k]
+				t.Run(k, func(t *testing.T) {
 					server := testCase.serverFunc(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 						if testCase.shouldError {
 							if e, a := httpstream.HeaderUpgrade, req.Header.Get(httpstream.HeaderConnection); e != a {
@@ -379,15 +382,16 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						defer proxy.Close()
 					}
 
+					interceptor := &Interceptor{proxyCalledWithHost: &proxyCalledWithHost}
 					if testCase.socks5ProxyServerFunc != nil {
 						var l net.Listener
 						var proxyHandler *socks5.Server
 						if testCase.proxyAuth != nil {
 							proxyHandler = testCase.socks5ProxyServerFunc(&socks5.StaticCredentials{
 								"proxyuser": "proxypasswd",
-						})
+							}, interceptor)
 						} else {
-						proxyHandler = socks5Server(nil)
+							proxyHandler = testCase.socks5ProxyServerFunc(nil, interceptor)
 						}
 
 						closed := false
@@ -400,8 +404,6 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 							l.Close()
 						}()
 
-					proxyCalledWithHost = "//127.0.0.1:" + strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
-
 						go func() {
 							for {
 								conn, err := l.Accept()
@@ -422,13 +424,11 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 
 						spdyTransport.proxier = func(proxierReq *http.Request) (*url.URL, error) {
 							proxierCalled = true
-						proxyURL, err := url.Parse(proxyCalledWithHost)
-						if err != nil {
-							return nil, err
-						}
-						proxyURL.User = testCase.proxyAuth
-						proxyURL.Scheme = "socks5"
-						return proxyURL, nil
+							return &url.URL{
+								Scheme: "socks5",
+								Host:   net.JoinHostPort("127.0.0.1", strconv.Itoa(l.Addr().(*net.TCPAddr).Port)),
+								User:   testCase.proxyAuth,
+							}, nil
 						}
 					}
 
@@ -444,7 +444,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						t.Fatalf("%s: shouldError=%t, got %t: %v", k, e, a, err)
 					}
 					if testCase.shouldError {
-					continue
+						return
 					}
 					defer conn.Close()
 
@@ -477,7 +477,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 						t.Fatalf("%s: expected '%s', got '%s'", k, e, a)
 					}
 
-				if testCase.proxyServerFunc != nil {
+					if testCase.proxyServerFunc != nil || testCase.socks5ProxyServerFunc != nil {
 						if !proxierCalled {
 							t.Fatalf("%s: Expected to use a proxy but proxier in SpdyRoundTripper wasn't called", k)
 						}
@@ -513,6 +513,7 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 							t.Fatalf("%s: Expected to see a call to the proxy with credentials %q, got %q", k, testCase.proxyAuth, proxyCalledWithAuthHeader)
 						}
 					}
+				})
 			}
 		})
 	}
```

Reply:
Thanks for the detailed review! I will address these comments and come back with a code update.

Reply:
I'm currently stuck trying to figure out how to access the interceptor methods outside of the `if testCase.socks5ProxyServerFunc != nil {` closure. Any hints you could provide?

Reply:
Based on the review comments, I assume that no changes were needed to the final set of test checks (lines 495-515). I am unsure if your intention was for this section to be removed entirely or not. Because no removal suggestions were included above, I assume that this logic should remain.

Reply:
notice in my example the `interceptor` variable is a concrete type:

`interceptor := &Interceptor{proxyCalledWithHost: &proxyCalledWithHost}`

you should be able to access the member fields of the concrete interceptor test implementation during the verification step